docs/DesignDocs/InitTimeConfigurationInterface.md

```diff
@@ -0,0 +1,271 @@
+ * @param[in] runtime_custom_claims_buffer The optional runtime custom claims
+ * buffer. When provided, the content of the buffer is included in the
+ * evidence_buffer, with integrity protection. Depending on the underlining TEE
+ * and the plugin implementation, the content might or might not be further
+ * encrypted.
+ * @param[in] runtime_custom_claims_buffer_size The number of bytes in the
+ * runtime custom claims buffer.
+ * @param[in] inittime_custom_claims_buffer The optional inittime custom claims
+ * buffer. When provided, the content of the buffer is included in the
+ * evidence_buffer as plaintext. The integrity protection of the content is
+ * provided by the underlining TEE and the enclave SW outside the plugin.
+ * @param[in] inittime_custom_claims_buffer_size The number of bytes in the
+ * inittime custom claims buffer.
+ * @param[in] optional_parameters The optional format-specific input parameters.
+ * @param[in] optional_parameters_size The size of optional_parameters in bytes.
+ * @param[out] evidence_buffer An output pointer that will be assigned the
+ * address of the dynamically allocated evidence buffer.
+ * @param[out] evidence_buffer_size A pointer that points to the size of the
+ * evidence buffer in bytes.
+ * @param[out] endorsements_buffer If not NULL, an output pointer that will be
+ * assigned the address of the dynamically allocated endorsements buffer.
+ * @param[out] endorsements_buffer_size A pointer that points to the size of the
+ * endorsements buffer in bytes.
+ * @retval OE_OK The function succeeded.
+ * @retval OE_INVALID_PARAMETER At least one of the parameters is invalid.
+ * @retval OE_NOT_FOUND The input evidence format id is not supported.
+ * @retval other appropriate error code.
+ */
+oe_result_t oe_get_evidence(
+    const oe_uuid_t* format_id,
+    uint32_t flags,
+    const void* runtime_custom_claims_buffer,
+    size_t runtime_custom_claims_buffer_size,
+    const oe_inittime_custom_claim_buffer_t* inittime_custom_claims_buffer,
+    size_t inittime_custom_claims_buffer_size,
+    const void* optional_parameters,
+    size_t optional_parameters_size,
+    uint8_t** evidence_buffer,
+    size_t* evidence_buffer_size,
+    uint8_t** endorsements_buffer,
+    size_t* endorsements_buffer_size);
+
+typedef struct _oe_inittime_custom_claim_buffer_t
+{
+    /* ID of integrity algorithm to be used to verify the buffer content.
+     *   0: config_id = SHA256(buffer[]);
+     *   others: undefined;
+     */
+    uint32_t integrity_algorithm_id;
+    /* Variable length buffer */
+    char buffer[];
+} oe_inittime_custom_claim_buffer_t;
+```
+
+Alternative Design
+-------------------
+
+Alternatively, the responsibility of "init-time" claims packaging and
+verification can be removed from the Attester/Verifier plugin and left to the
+upper level SW to implement. The upper level SW can append the "init-time" claim
+buffer to the evidence buffer returned by `oe_get_evidence` and send to the the
```

Comment:
A second alternative is for the attester enclave that calls the `oe_get_evidence()` to embed its init-time  claims as part of the custom claims. So the output evidence buffer would include the init-time claims with integrity protection. The verifier application, upon successful `oe_verify_evidence()`, would get  `oe_sgx_config_data_t` in a separate claim and `oe_enclave_inittime_configuration_data_t` as part of the custom claims, and is responsible to verify their relationship.

Reply:
Alternative 1 and 2 basically keep current API intact. It's up to the attester and verifier to define the protocol and payload format. Code snippets:

**Alternative 1**

```
///
/// Attester
///

const char extra_content[] = "1234567890";
size_t extra_content_size = sizeof(extra_content);
// Set up config id
...
// Create enclave
...
// Set up init-time claims
...

// Generate evidence
oe_get_evidence(...,
    evidence_buffer,
    &evidence_buffer_size,
    ...);

// Construct payload for evidence + extra content
p = payload = malloc(sizeof(extra_content_size) + evidence_buffer_size + sizeof(extra_content));
memcpy(p, &extra_content_size, sizeof(extra_content_size));
p += sizeof(extra_content_size);
memcpy(p, evidence_buffer, evidence_buffer_size);
p += evidence_buffer_size;
memcpy(p, extra_content, sizeof(extra_content));

// Send payload to verifier
send(payload);



///
/// Verifier
///

// Receive payload
receive(&payload, &payload_size);

// Extract payload
size_t extra_content_size;
memcpy(&extra_content_size, p, sizeof(extra_content_size));
void* evidence_buffer = payload + sizeof(extra_content_size);
uint32_t evidence_buffer_size = payload_size - sizeof(extra_content_size) - extra_content_size;
void* extra_content = payload + payload_size - extra_content_size;


// Verify evidence
oe_verify_evidence(...,
    evidence_buffer,
    evidence_buffer_size,
    ...,
    &claims,
    &claims_length);

// Verifier has to validate extra content
config_id = _find_claim(claims, claims_length, OE_SGX_CONFIG_ID);
if (SHA256(extra_content) != config_id)
    return error;
```

**Alternative 2**

```
///
/// Attester
///

const char extra_content[] = "1234567890";
size_t extra_content_size = sizeof(extra_content);
// Set up config id
...
// Create enclave
...
// Set up init-time claims
...

// Combine embed extra content in custom claims
oe_claim_t custom_claims[2];


oe_result_t oe_serialize_custom_claims(
    const oe_claim_t* custom_claims,
    size_t custom_claims_length,
    uint8_t** claims_out,
    size_t* claims_size_out)

// Generate evidence
oe_get_evidence(...,
    evidence_buffer,
    &evidence_buffer_size,
    ...);

// Construct payload for evidence + extra content
p = payload = malloc(sizeof(extra_content_size) + evidence_buffer_size + sizeof(extra_content));
memcpy(p, &extra_content_size, sizeof(extra_content_size));
p += sizeof(extra_content_size);
memcpy(p, evidence_buffer, evidence_buffer_size);
p += evidence_buffer_size;
memcpy(p, extra_content, sizeof(extra_content));

// Send payload to verifier
send(payload);



///
/// Verifier
///

// Receive payload
receive(&payload, &payload_size);

// Extract payload
size_t extra_content_size;
memcpy(&extra_content_size, p, sizeof(extra_content_size));
void* evidence_buffer = payload + sizeof(extra_content_size);
uint32_t evidence_buffer_size = payload_size - sizeof(extra_content_size) - extra_content_size;
void* extra_content = payload + payload_size - extra_content_size;


// Verify evidence
oe_verify_evidence(...,
    evidence_buffer,
    evidence_buffer_size,
    ...,
    &claims,
    &claims_length);

// Verifier has to validate extra content
config_id = _find_claim(claims, claims_length, OE_SGX_CONFIG_ID);
if (SHA256(extra_content) != config_id)
    return error;

```

Reply:
pseudocode now included in the latest version of the PR.